explainer/index.bs

```diff
@@ -178,6 +191,312 @@ Unlike invalidity, destroyed-ness can change after creation, is not contagious,
+Instead, for the CPU to see the content of a GPU buffer, it must be "mapped", making it available in the virtual memory space of the application (think of mapped as in `mmap()`).
+GPUBuffers must be specially allocated in order to be mappable - this can make it less efficient to access from the GPU (for example if it needs to be allocate in RAM instead of VRAM).
+
+All this discussion was centered around native GPU APIs, but in browsers, the GPU driver is loaded in the _GPU process_, so native GPU buffers can be mapped only in the GPU process's virtual memory.
+In general, it is not possible to map the buffer directly inside the _content process_ (though some systems can do this, providing optional optimizations).
+To work with this architecture an extra "staging" allocation is needed in shared memory between the GPU process and the content process.
+
+The table below recapitulates which type of memory is visible where:
+
+<table class="data">
+    <thead>
+        <tr>
+            <th>
+            <th> Regular `ArrayBuffer`
+            <th> Shared Memory
+            <th> Mappable GPU buffer
+            <th> Non-mappable GPU buffer (or texture)
+        </tr>
+    </thead>
+    <tr>
+        <td> CPU, in the content process
+        <td> **Visible**
+        <td> **Visible**
+        <td> Not visible
+        <td> Not visible
+    <tr>
+        <td> CPU, in the GPU process
+        <td> Not visible
+        <td> **Visible**
+        <td> **Visible**
+        <td> Not visible
+    <tr>
+        <td> GPU
+        <td> Not visible
+        <td> Not visible
+        <td> **Visible**
+        <td> **Visible**
+</table>
+
+### CPU-GPU Ownership Transfer ### {#buffer-mapping-ownership}
+
+In native GPU APIs, when a buffer is mapped, its content becomes accessible to the CPU.
+At the same time the GPU can keep using the buffer's content, which can lead to data races between the CPU and the GPU.
+This means that the usage of mapped buffer is simple but leaves the synchronization to the application.
+
+On the contrary, WebGPU prevents almost all data races in the interest of portability and consistency.
+In WebGPU there is even more risk of non-portability with races on mapped buffers because of the additional "shared memory" step that may be necessary on some drivers.
+That's why `GPUBuffer` mapping is done as an ownership transfer between the CPU and the GPU.
+At each instant, only one of the two can access it, so no race is possible.
+
+When an application requests to map a buffer, it initiates a transfer of the buffer's ownership to the CPU.
+At this time, the GPU may still need to finish executing some operations that use the buffer, so the transfer doesn't complete until all previously-enqueued GPU operations are finished.
+That's why mapping a buffer is an asynchronous operation (we'll discuss the other arguments below):
+
+<script type=idl>
```

Comment:
Might be able to use `<pre highlight=webidl>` to prevent it from being parsed as actual definitions? Not sure.

Reply:
That didn't work. The IDL compiler still complained:

> bikeshed --die-on=everything spec index.bs
> FATAL ERROR: IDL SYNTAX ERROR LINE: 8 - skipped: "Promise mapAsync(GPUMapModeFlags mode,
>                              optional GPUSize64 offset = 0,
>                              optional GPUSize64 size)"
> ✘  Did not generate, due to fatal errors
> make: *** [index.html] Error 2

Reply:
Just remembered there is a bikeshed feature for this
https://tabatkins.github.io/bikeshed/#no-idl
I'll try it and apply it myself.

Reply:
Took a few tries to find an invocation Bikeshed was happy with but got it.

Reply:
Actually `pre highlight=idl` or `xmp highlight=idl` does work for me, just pushed another change for that. Unfortunately changes the blocks from blue to yellow but also prevents it from generating a nonsense IDL index at the bottom of the document.

Reply:
(Problem with `pre` is it requires `<` to be escaped)

Reply:
The use of xmp looks like it works! Thanks for iterating on that.